### ISSUE TITLE
CMake build on macOS at least fails when using system ZLIB shared installed in CMAKE_PREFIX_PATH and FLTK_BUILD_TESTS is on.

### ISSUE THREAD

**Issue by ggarra13** [fltk/fltk#1181]
**Describe the bug**
Building on macOS amd64, build works fine with system zlib shared installed in CMAKE_PREFIX_PATH.
However, trying to build the tests fails, as they look for libz.dylib in bin/test and not in CMAKE_PREFIX_PATH.

**To Reproduce**
Attached are the logs of the build, showing where each library is found.

[compile.log](https://github.com/user-attachments/files/18251906/compile.log)

cmake scripts used:

[cmake.zip](https://github.com/user-attachments/files/18251849/cmake.zip)

**Expected behavior**
Tests should build and link properly to system zlib in CMAKE_PREFIX_PATH.

**Screenshots**

**FLTK Version**
1.4.1

**FLTK Configure / Build Options**
See logs.

**Operating System / Platform:**
macOS

**Additional context**

**Comment by ManoloFLTK** [comment]
No such problem here under macOS.
First, I build and install `zlib` and `libpng` with prefix `/usr/local`.
Then I build FLTK 1.4.1 with
```
cd build
cmake .. -DCMAKE_PREFIX_PATH=/usr/local -DFLTK_USE_SYSTEM_ZLIB=1 -DFLTK_USE_SYSTEM_LIBPNG=1
make
```
I see no problem to build test apps.
However, if FLTK was previously installed, then cmake would use the installed FL/ which may create problems if it's not the same version as the version in use. In that case, it's better to unistall FLTK befaore attempting to build the new FLTK.

**Comment by ggarra13** [comment]
Find an attached self compiling example.  It builds zlib and png as dependencies and then builds FLTK.

Just run the runme.sh script.

[broken_build.zip](https://github.com/user-attachments/files/18259358/broken_build.zip)

**Comment by ManoloFLTK** [comment]
I tried with this command closer to yours
`cmake .. -DCMAKE_PREFIX_PATH=/usr/local -DFLTK_USE_SYSTEM_ZLIB=1 -DFLTK_USE_SYSTEM_LIBPNG=1 -G Ninja -DFLTK_BUILD_SHARED_LIBS=1`
Still no build problem.

I'm worried by all these settings that are not those given in README.CMake
```
 -D BUILD_FLTK=ON -D BUILD_PNG=ON -D BUILD_ZLIB=ON -D BUILD_X11=ON -D BUILD_WAYLAND=ON -D FLTK_BUILD_SHARED=ON
```
I'm also worried by this setting without value `-D CMAKE_OSX_ARCHITECTURES=`

**Comment by ggarra13** [comment]
Please use the included runme.sh that builds ZLIB from source instead of relying on the system one.

I've simplified the build scripts removing all cruft.  

[broken_build.zip](https://github.com/user-attachments/files/18262671/broken_build.zip)

**Comment by ggarra13** [comment]
This is the type of error I get on macOS Intel:

```
/bin/sh: line 1:  8258 Abort trap: 6           /Users/gga/code/bugs/fltk_build/build/deps/FLTK/src/FLTK-build/bin/fluid.app/Contents/MacOS/fluid -c /Users/gga/code/bugs/fltk_build/build/deps/FLTK/src/FLTK/test/preferences.fl
[31/211] Generating fast_slow.cxx, fast_slow.h
FAILED: test/fast_slow.cxx test/fast_slow.h /Users/gga/code/bugs/fltk_build/build/deps/FLTK/src/FLTK-build/test/fast_slow.cxx /Users/gga/code/bugs/fltk_build/build/deps/FLTK/src/FLTK-build/test/fast_slow.h 
cd /Users/gga/code/bugs/fltk_build/build/deps/FLTK/src/FLTK-build/test && /Users/gga/code/bugs/fltk_build/build/deps/FLTK/src/FLTK-build/bin/fluid.app/Contents/MacOS/fluid -c /Users/gga/code/bugs/fltk_build/build/deps/FLTK/src/FLTK/test/fast_slow.fl
```

**Comment by MatthiasWM** [comment]
The last message looks more like a crash within Fluid, and not so much related to zlib? Is this still part of the original post?

**Comment by ggarra13** [comment]
Yes.  Building all demos fail due to not finding zlib in the path.  There's indeed a fluid crash as part of this.

**Comment by MatthiasWM** [comment]
I am confused by the way you build FLTK. Fluid does not fail at build time, but at run time when generating files. This should not happen because Fluid (just like all the test apps) link with libz statically. There should be no need to search for libz.1.dylib . It seems that you do not provide a static version of libz in your custom build. And if it is dynamically linked, libz.dylib must be available somewhere where Fluid can find it, i.e. the standard library search paths or in `Fluid.app/Contents/Resources/`.

**Comment by ggarra13** [comment]
@MatthiasWM I think you have found the potential flaw in the fluid build.  I currently build ZLIB dynamically in CMAKE_PREFIX_PATH/CMAKE_INSTALL_PATH (they are both set to the same to keep everything in a sandbox) and set the FLTK shared flags to ZLIB system one (which picks the one in the sandbox).  When FLTK builds fluid, it is likely just specifying libz which picks the dylib version of it, and does not use install_tool (or whatever is called) to set \@rpath or similar (I am not too familiar how these work in macOS).  I will try these two things: 
1) Building ZLIB both statically and dynamically.
2) Set DYLD_FALLBACK_LIBRARY_PATH to CMAKE_INSTALL_PATH/lib.

**Comment by dannye** [comment]
Hopefully this is of some help.

In order to statically link to a "system" zlib on macOS, I found that these 4 CMake arguments are required:

```
cmake \
	-D ... \
	-D FLTK_USE_SYSTEM_ZLIB=ON \
	-D ZLIB_INCLUDE_DIR="$(PKG_CONFIG_PATH=/opt/homebrew/opt/zlib/lib/pkgconfig pkg-config --cflags-only-I zlib | cut -c 3-)" \
	-D ZLIB_LIBRARY_RELEASE="$(PKG_CONFIG_PATH=/opt/homebrew/opt/zlib/lib/pkgconfig pkg-config --static --libs-only-L zlib | cut -c 3-)/libz.a" \
	-D LIB_zlib="$(PKG_CONFIG_PATH=/opt/homebrew/opt/zlib/lib/pkgconfig pkg-config --static --libs-only-L zlib | cut -c 3-)/libz.a"
```

Related thread: https://groups.google.com/g/fltkgeneral/c/kKT0dw1kG3M

Without all these options, CMake was unable to find the static zlib on its own, and always stuck to the .dylib instead.

`ZLIB_INCLUDE_DIR` is set to the path of the header files.
`ZLIB_LIBRARY_RELEASE` and `LIB_zlib` are both set to the exact same thing; the path to libz.a.

Maybe it's possible to do it more simply, but this is the only thing I have found that works.

**Comment by ggarra13** [comment]
Here's the full error:

```
dyld[4872]: Library not loaded: libz.1.dylib
  Referenced from: <61CE29D8-2B67-3D8A-8AA1-A4F15A3C7237> /Users/gga/code/bugs/fltk_build/build/deps/FLTK/src/FLTK-build/bin/fluid.app/Contents/MacOS/fluid
  Reason: tried: 'libz.1.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibz.1.dylib' (no such file), 'libz.1.dylib' (no such file), '/Users/gga/code/bugs/fltk_build/build/deps/FLTK/src/FLTK-build/test/libz.1.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/gga/code/bugs/fltk_build/build/deps/FLTK/src/FLTK-build/test/libz.1.dylib' (no such file), '/Users/gga/code/bugs/fltk_build/build/deps/FLTK/src/FLTK-build/test/libz.1.dylib' (no such file)
```

**Comment by ggarra13** [comment]
```
$ otool -L ./deps/FLTK/src/FLTK-build/bin/fluid.app/Contents/MacOS/fluid 
./deps/FLTK/src/FLTK-build/bin/fluid.app/Contents/MacOS/fluid:
	/System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa (compatibility version 1.0.0, current version 24.0.0)
	/System/Library/Frameworks/UniformTypeIdentifiers.framework/Versions/A/UniformTypeIdentifiers (compatibility version 1.0.0, current version 709.0.0, weak)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1336.0.0)
	@rpath/png.framework/Versions/1.6.43/png (compatibility version 0.0.0, current version 0.0.0)
	libz.1.dylib (compatibility version 1.0.0, current version 1.2.13)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1600.151.0)
	/System/Library/Frameworks/AppKit.framework/Versions/C/AppKit (compatibility version 45.0.0, current version 2483.0.0)
	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/ApplicationServices (compatibility version 1.0.0, current version 64.0.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 2048.1.255)
	/System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics (compatibility version 64.0.0, current version 1774.0.4)
	/System/Library/Frameworks/CoreText.framework/Versions/A/CoreText (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation (compatibility version 300.0.0, current version 2048.1.255)
	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
```

**Comment by ggarra13** [comment]
Setting CMAKE_BUILD_RPATH did not help.

**Comment by Albrecht-S** [comment]
Thanks for your additional information. I took a quick look but since it doesn't help ... Note that you can retrieve the actual RPATH setting on macOS with `otool -l` (lowercase L).

Meanwhile I tested your project on macOS and I found the issue: not setting `RPATH` in the `ZLIB` build because the statement
```
cmake_minimum_required(VERSION 2.4.4)
```
references an **_ancient CMake version_** which has some not expected consequences. However, CMake warns about the issues. _It's always good to read CMake warnings_ (excerpt, numbers added by me):
```
[6/16] Performing configure step for 'ZLIB'

(1) -----------------------------------------------------------------------

CMake Deprecation Warning at CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- The C compiler identification is AppleClang 16.0.0.16000026

(2) -----------------------------------------------------------------------

-- Renaming
--     /Users/albrecht/fltk/projects/is-1181/build/deps/ZLIB/src/ZLIB/zconf.h
-- to 'zconf.h.included' because this file is included with zlib
-- but CMake generates it automatically in the build directory.
-- Configuring done (1.0s)

(3) -----------------------------------------------------------------------

CMake Warning (dev):
  Policy CMP0042 is not set: MACOSX_RPATH is enabled by default.  Run "cmake
  --help-policy CMP0042" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  MACOSX_RPATH is not specified for the following targets:

   zlib

This warning is for project developers.  Use -Wno-dev to suppress it.
```

1. This warning seems benign but it implies warning 3 which is essential
2. This warning is indeed benign but you could avoid it by NOT "patching" `cmake/patches/ZLIB-patch/zconf.h.cmakein`. Or maybe not - I didn't test this.
3. The last warning is essential because "`MACOSX_RPATH` is not specified for ... zlib".

The 3rd warning causes fluid to fail. I used this simple patch to fix warnings (1) and (3):
```diff
--- a/cmake/patches/ZLIB-patch/CMakeLists.txt
+++ b/cmake/patches/ZLIB-patch/CMakeLists.txt
@@ -1,4 +1,5 @@
-cmake_minimum_required(VERSION 2.4.4)
+cmake_minimum_required(VERSION 3.15)
+
 set(CMAKE_ALLOW_LOOSE_LOOP_CONSTRUCTS ON)
 
 project(zlib C)
```
... and the project built successfully. You may likely use any minimal `CMake` version >= 3.0 (other versions not tested by me).

There are very likely other options to fix the issue but there's IMHO nothing we can do in FLTK.

- You could try to enable `CMP0042` by your build (I'm not sure if this is possible w/o patching zlib's `CMakeLists.txt`). It would probably be helpful to request an upstream patch: `CMake` 2.4.4 has been released on Nov 20, 2006 and `CMake` 3.0.0 has been released on June 9, 2014, more than 10 years ago.
- You could disable shared lib builds for `zlib` - which your build enables explicitly.

Anyway, I consider this issue fixed. Please close it if you agree.

Happy new year!

**Comment by ggarra13** [comment]
> Anyway, I consider this issue fixed. Please close it if you agree.

I do consider the issue fixed.  Thanks for looking into the issue so thoroughly.
